### ISSUE TITLE
Failing to build python bindings

### ISSUE THREAD

**Issue by SomeoneSerge** [tdegeus/cppcolormap#37]
Hi! Built the tests fine, but gcc doesn't seem to like the signatures in the python bindings

```
error: invalid 'static_cast' from type '<unresolved overloaded function type>' to type 'std::string (*)(const xt::pytensor<double, 1, xt::layout_type::dynamic>&)' {aka 'std::__cxx11::basic_string<char> (*)(const xt::pytensor<dou
       │ ble, 1, xt::layout_type::dynamic>&)'}
  69   │    61 |         static_cast<std::string (*)(const xt::pytensor<double, 1>&)>(&cppcolormap::rgb2hex),
```

Logs: https://gist.github.com/SomeoneSerge/2507244e5bc3b11982948ff0ed86a681
Getting the same error with gcc11 and gcc12.
Pybind11 version: pybind11-2.10.3/include (found version "2.10.3")
Built with: https://github.com/SomeoneSerge/pkgs/commit/b0274d7330c7819b174d2b707c4bb7f39270afd4

**Comment by tdegeus** [comment]
Thanks! I tried debugging here https://github.com/tdegeus/cppcolormap/pull/38 but unfortunately the CI works fine on Linux GCC 11.3.0. This makes it a bit harder to understand what is bugging the compiler and to solve it. It seems that the compiler has a hard time deciding between 
```cpp
template <class T, typename std::enable_if_t<xt::get_rank<T>::value != 1, int> = 0>
std::vector<std::string> rgb2hex(const T& arg);

template <class T, typename std::enable_if_t<xt::get_rank<T>::value == 1, int> = 0>
std::string rgb2hex(const T& arg);
```
Could you tried by adding the template explicitly to `main.cpp`:
```cpp
    m.def(
        "rgb2hex",
        static_cast<std::string (*)(const xt::pytensor<double, 1>&)>(&cppcolormap::rgb2hex<xt::pytensor<double, 1>>),
        DOC("rgb2hex")
    );

    m.def(
        "rgb2hex",
        static_cast<std::vector<std::string> (*)(const xt::pytensor<double, 2>&)>(
            &cppcolormap::rgb2hex<xt::pytensor<double, 2>>
        ),
        DOC("rgb2hex")
    );
```
?
Thanks!

**Comment by SomeoneSerge** [comment]
Quick update: still getting the same error at the first overload, i.e. one that returns `std::string` rather than a vector. The second one compiles. By the way, shouldn't the first one also consume a pytensor? Like `pytensor<double, 1>`. To be precise, here's the full diff I tried: https://github.com/SomeoneSerge/cppcolormap/pull/1/files

TODO: I should compare xtensor and pybind11 versions, and maybe set up a separate gh action to reproduce the failures more easily

**Comment by tdegeus** [comment]
Thanks for the update. That help to narrow it down. If I read better it seems that the compiler converts the `std::string` to `` and then is naturally confused. Maybe https://github.com/tdegeus/cppcolormap/pull/41 helps? Otherwise this might be a useful reference https://stackoverflow.com/questions/40705852/undefined-reference-to-std-cxx11basic-string-in-boost-on-travis-ci .

**Comment by SomeoneSerge** [comment]
No, that doesn't help yet, and I don't think it's got anything to do with cxx11 abi: the vector of strings seems to be fine.
By the way, I get this warning from ccls:

```
address of overloaded function 'rgb2hex' cannot be static_cast to type 'std::string (*)(const xt::pytensor<double, 1> &)' (aka 'basic_string<char> (*)(const pytensor<double, 1> &)')
```

**Comment by tdegeus** [comment]
Just to be sure are you on xtensor-python >=0.25.2 ?

**Comment by SomeoneSerge** [comment]
`0.25.1`

**Comment by tdegeus** [comment]
Then problem solved. `xt::pytensor` only has the static `rank` property since 0.25.2 . So pre-0.25.2 the compiler throws out 

https://github.com/tdegeus/cppcolormap/blob/2fe9454fe971785000d7521ee32b49bbba950c39/include/cppcolormap.h#L245

**Comment by tdegeus** [comment]
For future reference https://github.com/tdegeus/cppcolormap/commit/2fe9454fe971785000d7521ee32b49bbba950c39

**Comment by SomeoneSerge** [comment]
Perfect, thanks! Maybe also a `VERSION` in `find_package`?

**Comment by tdegeus** [comment]
> Perfect, thanks! Maybe also a `VERSION` in `find_package`?

Sounds great, a PR is welcome

**Comment by SomeoneSerge** [comment]
`xtensor-python` `0.26.1`, works as a charm

**Comment by tdegeus** [comment]
xtensor*'s versioning is a mystery to me ;)
